### ISSUE TITLE
:hover and pseudo elements in vnode style

### ISSUE THREAD

**Issue by staltz** [Matt-Esch/virtual-dom#179]
I'm building an app with virtual-dom, and going 100% styles in JavaScript rather than in CSS.

How do I express `:hover` state and `:before`/`:after` pseudo elements in virtual-hyperscript?

For instance

```css
.thing > a {
  color: #FF0000;
  &:hover {
    color: #FF00FF;
  }
  &:hover:after {
    content: "\276F";
    display: inline-block;
    position: absolute;
    right: 5px;
  }
}
```

**Comment by Zolmeister** [comment]
Doesn't look like an easy solution.
http://stackoverflow.com/questions/2754546/can-i-disable-a-css-hover-effect-via-javascript

Check out [Zorium](https://github.com/Zorium/zorium) - [Paper Button](https://github.com/Zorium/zorium-paper/blob/master/button/index.coffee) for how we do styles

**Comment by Raynos** [comment]
@staltz A very good question. Let's take a look at RCSS.

[rcss](https://github.com/chenglou/RCSS) uses a javascript DSL to create class names which means you do `h('div', { className: style.foo })`

The other approach, which is based on computed styles will involve re-implementing large parts of CSS. You are just going to have to re-implement it. Thats what [facebook will do](https://github.com/reactjs/react-future/issues/8#issuecomment-44224086)

**Comment by staltz** [comment]
> You are just going to have to re-implement it.

Wouldn't it be good to plan what parts of CSS are worth being implemented in virtual-dom? BTW, I'm a believer in the "All styles in JavaScript" long-term strategy, so that's why I think it should be in virtual-dom. There is already a good enough support for styles.

**Comment by staltz** [comment]
This is actually not the first complication with `<a>` tags that I find. My current understanding is that `<a>` tags are so old in the web, that it is difficult to abandon the old ways of using them, and that they were not originally designed as pieces of interactive user interface, but rather as marked up text, with pointers. As an analogy, HTML used to be what Markdown is today.

For instance, in Cycle.js, `<a href="url"/>` breaks the whole way of doing things in the framework. Because `href` specifies behavior, while Views in Cycle are meant to only describe looks. Also `:hover` in a way is just a shortcut for specifying behavior in user interaction. In Mercury, you would send an event on mouseover, the view state would lift a isHovered boolean as true, and then the rendering would display the element differently whether isHovered is true or false. In Cycle it's roughly the same. `:hover` breaks that because it creates this shortcut.

So, while `<a href>` and `:hover` are very convenient and simple to use, I end up having to make a custom element `<x-a>` which handles the hovering behavior internally, and I give attributes to it to specify styles for hovered and for normal. This has nothing to do specifically with virtual-dom, but what I want to say is that I don't think we need to implement `:hover` in virtual-dom, since it goes against the majority of the use cases for virtual-dom.

`:after` and `:before` are actually unnecessary in virtual-dom because you can just prepend and append divs to your target.

I haven't thought yet about media queries, but one of the most problematic pieces is `:visited` for `<a>` tags. As I said, browsers and the web have a long legacy. In the context of virtual-dom, it would be useful if we had something like `window.hasPageBeenVisited('http://google.com")` returning a boolean, so virtual-dom frameworks use that to set a view state boolean that views will consume. But that won't happen, also for privacy issues. We still need to resort to CSS just for `:visited`.

TL;DR: just a rant that browsers are not playing nice with virtual-dom frameworks, and that it might not be worth implementing CSS features for styling vtrees.
